docs/architecture/concentrated-liquidity-adr.md

```diff
@@ -659,7 +661,247 @@ TODO
+we can calculate the fee growth inside the range by subtracting the two from the
+global per-unit-of-liquidity fee growth.
+
+```go
+feeGrowthInsideRange0 := pool.FeeGrowthGlobalOutside.Token0 - feeGrowthBelowLowerTick0 - feeGrowthAboveUpperTick0
+```
+
+Note that although `tick.FeeGrowthOutside.Token0` may be initialized at a different
+point in time for each tick, the comparison of these values between ticks
+is not meaningful. There is also no guarantee that the values
+across ticks will follow any particular pattern. 
+
+However, this does not affect the per-position calculations since
+all the position  needs to know is the fee growth inside the position's
+range since the position was last touched.
+
+- **Per-position**
+
+type Position struct {
+    FeeGrowthInsideLast Fee
+    UncollectedFee Fee
+}
+
+Recall that contrary to traditional pools, in concentrated liquidity pool,
+fees do not get auto re-injected into the pool. Instead, they are tracked by
+`position.TokensUncollected0` and `position.TokensUncollected1` fields of each position.
+
+The `position.FeeGrowthInside0Last` and `position.FeeGrowthInside1Last` accumulators
+are used to calculate the  _uncollected fees_ to add to `position.TokensUncollected0`
+and `position.TokensUncollected1`.
+
+The amount of uncollected fees needs to be calculated every time a user touches
+the position. That is when a position is created, liqudiity is added or removed.
+
+We must recalculate the values for any modification because with more liqudity
+added to the position, the amount of fees collected by the position increases.
+
+Let `feeGrowthInside0` be the amount of fee growth per unit of liqudity within
+the position's ticks. We use the same strategy for computing fees between two ticks (in-range) that
+was described in the previous section. Once we have `feeGrowthInside0` computed, we update the
+`position.UncollectedFee.Token0` and `position.FeeGrowthInsideLast.Token0`.
+
+```go
+// Note that tokensUncollected0 is 0 when a position is created.
+if !isPositionNew {
+    uncollectefFeeAddition := pool.Liquidity.Mul(feeGrowthInside.Token0.Sub(position.FeeGrowthInsideLast.Token0))
+    position.UncollectedFee.Token0 = position.UncollectedFee.Token0.Add(uncollectefFeeAddition)
+}
+
+position.FeeGrowthInsideLast.Token0 = feeGrowthInside.Token0
+```
+
+##### Collecting Fees
+
+Collecting fees is as simple as transferring the requested amount
+from the pool address to the position's owner.
+
+```go
+func (k Keeper) collectFees(
+    owner sdk.AccAddress,
+    lowerTick, upperTick int64, 
+    amount0Requested, amount1Requested sdk.Int) error {
+    // validate ticks
+
+    // get position if exists
+
+    // make sure that amounts requested are not greater than
+    // position's uncollected fees.
+
+    // bank send position.TokensUncollected0 and position.TokensUncollected1
+    //from pool address to position owner
+    // TODO: revisit to make sure truncations are handled correctly.
+}
+```
```

Comment:
Random design choice question: should we give users the ability to specify how much of their share of fees they want to collect (like here) or should we just do an all or nothing collection? IMO, we should allow one collection every 24h (or just reset after epoch), and that collection should be their entire share. Otherwise the META is to just spam collections constantly.

Reply:
Good point. 

I will change this to:
- auto-collect full amount every epoch
- have the ability to manually collect full amounts (optional, we can implement later)

Reply:
After our discussion with Dev, I still think this point stands, just not auto-collect but instead allow users to collect once every 24h, so be reset at epoch